### PR TITLE
Feature: enable multi pool client connections

### PR DESCRIPTION
**Pull request by sinadarbouy**
# Ticket(s)
#398
Closes #512
## Description
This pull request introduces significant updates to enhance the handling of proxies within the server configuration and adds support for load balancing strategies. Below are the key changes and features included in this pull request:

#### Configuration Changes:

- Updated the configuration files (config.go, constants.go, types.go) to include new fields for multiple proxies and load balancer strategies.
- Added default values for new fields in config.go.
- Enhanced the global configuration validation to ensure proper referencing of clients, pools, and proxies.
- **Important:** The structure of the gatewayd.yaml configuration has changed to reflect the new support for multiple proxies and load balancer strategies. Specifically, the configuration now includes a loadBalancer section where rules and strategies are grouped together for clarity.

#### Load Balancer Implementation:
- Implemented loadbalancer.go to handle load balancing strategy selection.
- Introduced round-robin.go to implement the Round Robin load balancing strategy.
- Created tests for load balancer strategies (loadbalancer_test.go, round-robin_test.go).

#### Configuration Examples:
- Updated gatewayd.yaml to include examples of multiple proxies and load balancer configuration.
```yaml
# GatewayD Global Configuration

loggers:
  default:
...

metrics:
  default:
...

clients:
  default:
...
  default-2:
...

pools:
  default:
...
  default-2:
...

proxies:
  default:
...
  default-2:
...

servers:
  default:
    network: tcp
    address: 0.0.0.0:15432
    proxies:
      - "default"
      - "default-2"
    loadBalancer:
      strategy: ROUND_ROBIN
      # Not yet implemented.
      # loadBalancingRules:
      #   - condition: "default"
      #     percentages:
      #       - proxy: "default"
      #         percentage: 70
      #       - proxy: "default-2"
      #         percentage: 30

api:
...

```
## Development Checklist

- [X] I have added a descriptive title to this PR.
- [X] I have squashed related commits together.
- [X] I have rebased my branch on top of the latest main branch.
- [X] I have performed a self-review of my own code.
- [X] I have commented on my code, particularly in hard-to-understand areas.
- [X] I have added docstring(s) to my code.
- [ ] I have made corresponding changes to the documentation (docs).
- [ ] I have updated docs using `make gen-docs` command.
- [X] I have added tests for my changes.
- [X] I have signed all the commits.

## Legal Checklist

- [X] I have read the [CONTRIBUTING](https://github.com/gatewayd-io/gatewayd/blob/main/CONTRIBUTING.md) document.
- [X] I have read and understood the [Code of Conduct](https://github.com/gatewayd-io/gatewayd/blob/main/CODE_OF_CONDUCT.md).
- [X] I have read and agreed to the [Apache CLA](https://www.apache.org/licenses/contributor-agreements.html) (required).
- [X] I have read and agreed to the [LICENSE](https://github.com/gatewayd-io/gatewayd/blob/main/LICENSE) (required).

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -449,33 +449,33 @@ func (c *Config) ValidateGlobalConfig(ctx context.Context) *gerr.GatewayDError {
...
-			err := fmt.Errorf("\"proxies.%s\" is nil or empty", configGroup)
-			span.RecordError(err)
-			errors = append(errors, gerr.ErrValidationFailed.Wrap(err))
-		}
-	}
-
-	if len(globalConfig.Proxies) > 1 {
-		seenConfigObjects = append(seenConfigObjects, "proxies")
-	}
+	// if len(globalConfig.Clients) > 1 {
```
Comment (ccoVeille) [inline config/config.go:452]: This block of commented code looks like something forgotten

File: gatewayd.yaml [hunk 2]
```diff
@@ -3,7 +3,7 @@
 loggers:
   default:
     output: ["console"] # "stdout", "stderr", "syslog", "rsyslog" and "file"
-    level: "info" # panic, fatal, error, warn, info (default), debug, trace
+    level: "debug" # panic, fatal, error, warn, info (default), debug, trace
```
Comment (ccoVeille) [inline gatewayd.yaml:6]: Looks like a forgotten debug

File: gatewayd.yaml [hunk 3]
```diff
@@ -33,7 +33,22 @@ metrics:
...
+    address: localhost:5434
+    tcpKeepAlive: False
+    tcpKeepAlivePeriod: 30s # duration
+    receiveChunkSize: 8192
+    receiveDeadline: 0s # duration, 0ms/0s means no deadline
+    receiveTimeout: 0s # duration, 0ms/0s means no timeout
+    sendDeadline: 0s # duration, 0ms/0s means no deadline
+    dialTimeout: 60s # duration
+    # Retry configuration
+    retries: 3 # 0 means no retry and fail immediately on the first attempt
```
Comment (ccoVeille) [inline gatewayd.yaml:45]:
```suggestion
    retries: 3 # 0 means no retries, so it fails immediately on the first attempt
```

File: network/loadbalancer.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+	GetNextProxy() IProxy
+}
+
+type Strategy struct {
+	server *Server
+}
+
+func NewLoadBalancerStrategy(server *Server) (LoadBalancerStrategy, *gerr.GatewayDError) {
+	switch server.LoadbalancerStrategyName {
+	case "ROUND_ROBIN":
```
Comment (ccoVeille) [inline network/loadbalancer.go:17]: This should be a  package constant

File: network/server.go [hunk 5]
```diff
@@ -664,6 +694,12 @@ func NewServer(
 			"GatewayD is listening on an unresolved address")
 	}
 
+	st, err := NewLoadBalancerStrategy(&server)
+	if err != nil {
+		srv.Logger.Error().Msg("Failed to create a loadbalncer strategy")
```
Comment (ccoVeille) [inline network/server.go:699]:
```suggestion
		srv.Logger.Error().Msg("Failed to create a loadbalancer strategy")
```

File: network/round-robin.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
+package network
+
+import (
+	"sync/atomic"
+)
+
+type RoundRobin struct {
+	proxies []IProxy
+	next    uint32
```
Comment (ccoVeille) [inline network/round-robin.go:9]: You are using atomic the old way

You can use atomic.Int32

Please take a look at the documentation.

https://pkg.go.dev/sync/atomic#Int32

The code is way simpler since it was added

File: network/server.go [hunk 7]
```diff
@@ -673,3 +709,12 @@ func (s *Server) CountConnections() int {
 	defer s.mu.RUnlock()
 	return int(s.connections)
 }
+
+func (s *Server) GetProxyForConnection(conn *ConnWrapper) (IProxy, bool) {
```
Comment (ccoVeille) [inline network/server.go:713]: Nothing in your PR uses the exist value returned

File: config/config.go [hunk 8]
```diff
@@ -489,6 +477,24 @@ func (c *Config) ValidateGlobalConfig(ctx context.Context) *gerr.GatewayDError {
 		seenConfigObjects = append(seenConfigObjects, "servers")
 	}
 
+	// Each server configuration should have at least one proxy defined.
+	// Each proxy in the server configuration should be referenced in proxies configuration.
+	for serverName, server := range globalConfig.Servers {
+		if len(server.Proxies) == 0 {
+			err := fmt.Errorf("\"servers.%s\"  has no proxies defined", serverName)
```
Comment (ccoVeille) [inline config/config.go:484]: I know such code was already present, and you only copy-pasted + adapted,  but please consider doing this

```suggestion
			err := fmt.Errorf(`"servers.%s"  has no proxies defined`, serverName)
```

File: config/config.go [hunk 9]
```diff
@@ -489,6 +477,24 @@ func (c *Config) ValidateGlobalConfig(ctx context.Context) *gerr.GatewayDError {
...
+			span.RecordError(err)
+			errors = append(errors, gerr.ErrValidationFailed.Wrap(err))
+			continue
+		}
+		for _, proxyName := range server.Proxies {
+			if _, exists := c.globalDefaults.Proxies[proxyName]; !exists {
+				err := fmt.Errorf("\"servers.%s\" references a non-existent proxy \"%s\"", serverName, proxyName)
```
Comment (ccoVeille) [inline config/config.go:491]:
```suggestion
				err := fmt.Errorf(`"servers.%s" references a non-existent proxy "%s"`, serverName, proxyName)
```

File: network/round-robin.go (hunk 6 above)
Comment (sinadarbouy) [reply network/round-robin.go:9]: Thanks for the review. I’ll check it.

File: gatewayd.yaml (hunk 2 above)
Comment (sinadarbouy) [reply gatewayd.yaml:6]: yup, thanks for looking this over!
the PR is not ready yet. I just changed this YAML to show it as an example. Probably in the final version of the PR, I will change the YAML to the default mode with commented proxy definitions.

File: config/config.go [hunk 10]
```diff
@@ -489,6 +477,24 @@ func (c *Config) ValidateGlobalConfig(ctx context.Context) *gerr.GatewayDError {
 		seenConfigObjects = append(seenConfigObjects, "servers")
 	}
 
+	// Each server configuration should have at least one proxy defined.
+	// Each proxy in the server configuration should be referenced in proxies configuration.
+	for serverName, server := range globalConfig.Servers {
+		if len(server.Proxies) == 0 {
+			err := fmt.Errorf(`"servers.%s"  has no proxies defined`, serverName)
```
Comment (ccoVeille) [inline config/config.go:484]:
```suggestion
			err := fmt.Errorf(`"servers.%s" has no proxies defined`, serverName)
```

File: network/loadbalancer_test.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+	if err != nil {
+		t.Errorf("Expected no error, got %v", err)
+	}
+
+	_, ok := strategy.(*RoundRobin)
+	if !ok {
+		t.Errorf("Expected strategy to be of type RoundRobin")
+	}
+
+	// Test case 1: InValid strategy name
```
Comment (ccoVeille) [inline network/loadbalancer_test.go:26]:
```suggestion
	// Test case 2: InValid strategy name
```

No?

File: network/loadbalancer_test.go [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+	serverInvalid := &Server{
+		LoadbalancerStrategyName: "InvalidStrategy",
+		Proxies:                  []IProxy{MockProxy{}},
+	}
+
+	strategy, err = NewLoadBalancerStrategy(serverInvalid)
+	if err != gerr.ErrLoadBalancerStrategyNotFound {
```
Comment (ccoVeille) [inline network/loadbalancer_test.go:33]:
```suggestion
	if !errors.Is(err, gerr.ErrLoadBalancerStrategyNotFound) {
```

File: network/round-robin_test.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	server := &Server{Proxies: proxies}
+	rr := NewRoundRobin(server)
+
+	expectedOrder := []string{"proxy2", "proxy3", "proxy1", "proxy2", "proxy3"}
+
+	for i, expected := range expectedOrder {
+		proxy := rr.GetNextProxy()
+		if proxy.(MockProxy).GetName() != expected {
+			t.Errorf("test %d: expected proxy name %s, got %s", i, expected, proxy.(MockProxy).GetName())
```
Comment (ccoVeille) [inline network/round-robin_test.go:61]:
```suggestion
		p, ok := proxy.(MockProxy)
		if !ok {
			t.Errorf("test %d: expected proxy name %s", i, expected)
		}
		if p.GetName() != expected
			t.Errorf("test %d: expected proxy name %s, got %s", i, expected, p.GetName())
```

File: network/server.go [hunk 14]
```diff
@@ -673,3 +723,14 @@ func (s *Server) CountConnections() int {
 	defer s.mu.RUnlock()
 	return int(s.connections)
 }
+
+// GetProxyForConnection returns the proxy associated with the given connection.
+func (s *Server) GetProxyForConnection(conn *ConnWrapper) (IProxy, bool) {
+	proxy, exists := s.connectionToProxyMap[conn]
+	return proxy, exists
+}
```
Comment (ccoVeille) [inline network/server.go:731]: Nothing in your PR uses the exist value returned


```suggestion
func (s *Server) GetProxyForConnection(conn *ConnWrapper) IProxy {
	proxy, _ := s.connectionToProxyMap[conn]
	return proxy
}
```


@sinadarbouy

Comment (sinadarbouy) [reply network/server.go:731]: added for each reference in the lines above in `server.go` to handle logging and errors if the proxy is not found

File: network/loadbalancer.go (hunk 4 above)
Comment (sinadarbouy) [reply network/loadbalancer.go]: tnx for the review!

**Comment by ccoVeille** [comment]
LGTM 👍

File: network/round-robin.go [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+	return &RoundRobin{proxies: server.Proxies}
+}
+
+func (r *RoundRobin) GetNextProxy() IProxy {
+	proxiesLen := uint32(len(r.proxies))
+
+	nextIndex := r.next.Add(1)
+
+	return r.proxies[nextIndex%proxiesLen]
+}
```
Comment (likecodingloveproblems) [inline network/round-robin.go:20]: Wasn't it better to have a circular linked-list?

Comment (sinadarbouy) [reply network/round-robin.go:20]: Thanks for your review and suggestion! I saw your PR, and it's a really elegant solution. However, I went with an array index using sync/atomic because our proxy count is fixed. Indexing offers better performance and memory efficiency due to better cache locality and direct indexing, avoiding the pointer overhead of linked lists. Plus, sync/atomic gives us efficient thread-safe access, if we scale with more Goroutines later on.
those are my reasons for picking sync/atomic.

File: errors/errors.go [hunk 16]
```diff
@@ -53,6 +53,7 @@ const (
 	ErrCodeMsgEncodeError
 	ErrCodeConfigParseError
 	ErrCodePublishAsyncAction
+	ErrorCodeLoadBalancerStrategyNotFound
```
Comment (mostafa) [inline errors/errors.go:56]:
```suggestion
	ErrCodeLoadBalancerStrategyNotFound
```

File: errors/errors.go [hunk 17]
```diff
@@ -194,6 +195,10 @@ var (
 		ErrCodePublishAsyncAction, "error publishing async action", nil,
 	}
 
+	ErrLoadBalancerStrategyNotFound = &GatewayDError{
+		ErrorCodeLoadBalancerStrategyNotFound, "The specified load balancer strategy does not exist.", nil,
```
Comment (mostafa) [inline errors/errors.go:199]:
```suggestion
		ErrCodeLoadBalancerStrategyNotFound, "The specified load balancer strategy does not exist.", nil,
```

File: gatewayd.yaml [hunk 18]
```diff
@@ -46,19 +46,51 @@ clients:
...
+    loadBalancer:
+      strategy: ROUND_ROBIN
+      # Not yet implemented.
+      # loadBalancingRules:
+      #   - condition: "default"
+      #     percentages:
+      #       - proxy: "default"
+      #         percentage: 70
+      #       - proxy: "default-2"
+      #         percentage: 30
```
Comment (mostafa) [inline gatewayd.yaml:93]: Remove this for now.

File: network/constants.go [hunk 19]
```diff
@@ -0,0 +1,6 @@
+package network
+
+// Load balancing strategies.
+const (
+	RoundRobinStrategy = "ROUND_ROBIN"
+)
```
Comment (mostafa) [inline network/constants.go:6]: All constants should reside in `config/constants.go` and should be referenced from there.

File: network/loadbalancer.go [hunk 20]
```diff
@@ -0,0 +1,18 @@
+package network
+
+import (
+	gerr "github.com/gatewayd-io/gatewayd/errors"
+)
+
+type LoadBalancerStrategy interface {
+	GetNextProxy() IProxy
```
Comment (mostafa) [inline network/loadbalancer.go:8]: Nit: `Get` is not needed.

```suggestion
	NextProxy() IProxy
```

File: network/round-robin.go [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+type RoundRobin struct {
+	proxies []IProxy
+	next    atomic.Uint32
+}
+
+func NewRoundRobin(server *Server) *RoundRobin {
+	return &RoundRobin{proxies: server.Proxies}
+}
+
+func (r *RoundRobin) GetNextProxy() IProxy {
```
Comment (mostafa) [inline network/round-robin.go:14]: Nit: `NextProxy`?
```suggestion
func (r *RoundRobin) NextProxy() IProxy {
```

File: network/round-robin.go [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+	proxiesLen := uint32(len(r.proxies))
+
+	nextIndex := r.next.Add(1)
```
Comment (mostafa) [inline network/round-robin.go:17]: Maybe check for overflows before increasing this indefinitely?
```suggestion
    if r.next.Load() >= math.MaxUint32 {
        r.next.Store(0)
    }
	nextIndex := r.next.Add(1)
```

Nit: maybe remove empty lines as well.

File: network/round-robin.go [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+
+	return r.proxies[nextIndex%proxiesLen]
```
Comment (mostafa) [inline network/round-robin.go:19]: 👏

File: network/server.go [hunk 24]
```diff
@@ -149,10 +154,13 @@ func (s *Server) OnOpen(conn *ConnWrapper) ([]byte, Action) {
 	}
 	span.AddEvent("Ran the OnOpening hooks")
 
+	// get proxy
+	proxy := s.loadbalancerStrategy.GetNextProxy()
```
Comment (mostafa) [inline network/server.go:158]:
```suggestion
	proxy := s.loadbalancerStrategy.NextProxy()
```

File: network/server.go [hunk 25]
```diff
@@ -165,6 +173,9 @@ func (s *Server) OnOpen(conn *ConnWrapper) ([]byte, Action) {
 		return nil, None
 	}
 
+	// AssignConnectionToProxy
```
Comment (mostafa) [inline network/server.go:176]:
```suggestion
	// Assign connection to proxy
```

File: network/server.go [hunk 26]
```diff
@@ -303,7 +326,15 @@ func (s *Server) OnTraffic(conn *ConnWrapper, stopConnection chan struct{}) Acti
 	go func(server *Server, conn *ConnWrapper, stopConnection chan struct{}, stack *Stack) {
 		for {
 			server.Logger.Trace().Msg("Passing through traffic from client to server")
-			if err := server.Proxy.PassThroughToServer(conn, stack); err != nil {
+
+			// Find the proxy associated with the given connection
+			proxy, exists := server.GetProxyForConnection(conn)
+			if !exists {
+				server.Logger.Error().Msg("Failed to find proxy that matches the connection")
+				break
```
Comment (mostafa) [inline network/server.go:334]: Should we close the connection at this point? Just curious.

File: network/server.go [hunk 27]
```diff
@@ -317,7 +348,14 @@ func (s *Server) OnTraffic(conn *ConnWrapper, stopConnection chan struct{}) Acti
 	go func(server *Server, conn *ConnWrapper, stopConnection chan struct{}, stack *Stack) {
 		for {
 			server.Logger.Trace().Msg("Passing through traffic from server to client")
-			if err := server.Proxy.PassThroughToClient(conn, stack); err != nil {
+
+			// Find the proxy associated with the given connection
+			proxy, exists := server.GetProxyForConnection(conn)
+			if !exists {
+				server.Logger.Error().Msg("Failed to find proxy that matches the connection")
+				break
```
Comment (mostafa) [inline network/server.go:356]: Same as above.

File: network/round-robin_test.go [hunk 28]
```diff
@@ -0,0 +1,96 @@
...
+		MockProxy{name: "proxy2"},
+		MockProxy{name: "proxy3"},
+	}
+	server := &Server{Proxies: proxies}
+	roundRobin := NewRoundRobin(server)
+
+	expectedOrder := []string{"proxy2", "proxy3", "proxy1", "proxy2", "proxy3"}
+
+	for testIndex, expected := range expectedOrder {
+		proxy := roundRobin.GetNextProxy()
```
Comment (mostafa) [inline network/round-robin_test.go:60]:
```suggestion
		proxy := roundRobin.NextProxy()
```

**Review by mostafa — COMMENTED** [review]
Thanks for your awesome contribution! 🙏 This paves the way for further development.

I have left a few comments.

File: network/round-robin.go (hunk 22 above)
Comment (sinadarbouy) [reply network/round-robin.go:24]: the `sync/atomic` package can handle overflows by resetting the value to 0. To test this, added a new test case called `TestNextProxyOverflow`. also added error handling for `proxiesLen` and changed the return interface signature to `(IProxy, *gerr.GatewayDError)`

File: network/server.go (hunk 26 above)
Comment (sinadarbouy) [reply network/server.go]: The 'not exist' error is kinda panic error, so yes, probably we should close it. I’ve added `stopConnection <- struct{}{} `right after that.

File: network/server.go (hunk 27 above)
Comment (sinadarbouy) [reply network/server.go]: same added

File: network/constants.go (hunk 19 above)
Comment (sinadarbouy) [reply network/constants.go]: Moved in `config/constants.go`

File: network/round-robin.go [hunk 29]
```diff
@@ -0,0 +1,26 @@
+package network
```
Comment (ccoVeille) [inline network/round-robin.go:1]: Either roundrobin.go round_robin.go

Usually there is no hyphen in files names

File: network/round-robin_test.go [hunk 30]
```diff
@@ -0,0 +1,117 @@
+package network
```
Comment (ccoVeille) [inline network/round-robin_test.go:1]: Same remark about the test file

**Review by ccoVeille — COMMENTED** [review]
LGTM 🤞

**Comment by sinadarbouy** [comment]
recent changes I made to the structure configuration. We were having issues, especially when dealing with multiple servers. If a second server went down, no logs were being generated. This happened because the logger was expecting a second instance (default-2), and when it wasn’t there, the logger wouldn’t work, and we couldn’t see any logs. Adding a second logger instance made the errors visible. Also, the metrics system was looking for default-2 and throwing errors when it couldn’t find it.

To fix this, I changed the structure to match what was originally suggested in issue [#398](https://github.com/gatewayd-io/gatewayd/issues/398). Now, each server, client, pool, and proxy can have its own logger and metrics configuration, which should make things more reliable and scalable.

File: gatewayd.yaml [hunk 31]
```diff
@@ -32,33 +32,57 @@ metrics:
...
+      backoffMultiplier: 2.0 # 0 means no backoff
+      disableBackoffCaps: false
 
 pools:
   default:
-    size: 10
+    active-writes:
+      size: 10
+    standby-reads:
+      size: 10
```
Comment (ccoVeille) [inline gatewayd.yaml:71]:
```suggestion
    activeWrites:
      size: 10
    standbyReads:
      size: 10
```

Same as https://github.com/gatewayd-io/gatewayd/pull/577#discussion_r1694273227

File: gatewayd.yaml [hunk 32]
```diff
@@ -32,33 +32,57 @@ metrics:
...
+      receiveDeadline: 0s # duration, 0ms/0s means no deadline
+      receiveTimeout: 0s # duration, 0ms/0s means no timeout
+      sendDeadline: 0s # duration, 0ms/0s means no deadline
+      dialTimeout: 60s # duration
+      # Retry configuration
+      retries: 3 # 0 means no retry and fail immediately on the first attempt
+      backoff: 1s # duration
+      backoffMultiplier: 2.0 # 0 means no backoff
+      disableBackoffCaps: false
+    standby-reads:
```
Comment (ccoVeille) [inline gatewayd.yaml:50]:
```suggestion
    activeWrites:
      network: tcp
      address: localhost:5432
      tcpKeepAlive: False
      tcpKeepAlivePeriod: 30s # duration
      receiveChunkSize: 8192
      receiveDeadline: 0s # duration, 0ms/0s means no deadline
      receiveTimeout: 0s # duration, 0ms/0s means no timeout
      sendDeadline: 0s # duration, 0ms/0s means no deadline
      dialTimeout: 60s # duration
      # Retry configuration
      retries: 3 # 0 means no retry and fail immediately on the first attempt
      backoff: 1s # duration
      backoffMultiplier: 2.0 # 0 means no backoff
      disableBackoffCaps: false
    standbyReads:
```

Why using kebab case here, while almost everything else (if not all) was in camelCase

File: gatewayd.yaml [hunk 33]
```diff
@@ -32,33 +32,57 @@ metrics:
...
+      healthCheckPeriod: 60s # duration
+    standby-reads:
+      healthCheckPeriod: 60s # duration
 
 servers:
   default:
     network: tcp
     address: 0.0.0.0:15432
+    loadBalancer:
+      strategy: ROUND_ROBIN
```
Comment (ccoVeille) [inline gatewayd.yaml:85]: I wouldn't use Screaming Snake case here

**Review by ccoVeille — CHANGES_REQUESTED** [review]

File: gatewayd.yaml (hunk 33 above)
Comment (sinadarbouy) [reply gatewayd.yaml:85]: Trying to follow Istio’s structure for naming algorithms for load balancers. They are using SCREAMING_SNAKE_CASE for load balancer algorithms.

File: api/api.go [hunk 34]
```diff
@@ -201,16 +201,21 @@ func (a *API) GetPlugins(context.Context, *emptypb.Empty) (*v1.PluginConfigs, er
 }
 
 // GetPools returns the pool configuration of the GatewayD.
-func (a *API) GetPools(context.Context, *emptypb.Empty) (*structpb.Struct, error) {
-	_, span := otel.Tracer(config.TracerName).Start(a.ctx, "Get Pools")
+func (a *API) GetPools(ctx context.Context, _ *emptypb.Empty) (*structpb.Struct, error) {
+	_, span := otel.Tracer(config.TracerName).Start(ctx, "Get Pools")
```
Comment (mostafa) [inline api/api.go:205]: Do not use request context for tracing here. If you do, then change all instances and remove the `a.ctx` everywhere. To be fair, the whole tracing should be refactored to have different traces for different clients and plugins.

```suggestion
func (a *API) GetPools(context.Context, _ *emptypb.Empty) (*structpb.Struct, error) {
	_, span := otel.Tracer(config.TracerName).Start(a.ctx, "Get Pools")
```

File: api/api.go [hunk 35]
```diff
@@ -201,16 +201,21 @@ func (a *API) GetPlugins(context.Context, *emptypb.Empty) (*v1.PluginConfigs, er
...
 	defer span.End()
 
-	pools := make(map[string]interface{})
-	for name, p := range a.Pools {
-		pools[name] = map[string]interface{}{
-			"cap":  p.Cap(),
-			"size": p.Size(),
+	pools := make(map[string]any)
```
Comment (mostafa) [inline api/api.go:208]: Note to self: plugins should be updated as well, as they rely on this API.

File: api/api_test.go [hunk 36]
```diff
@@ -210,21 +210,26 @@ func TestGetPluginsWithEmptyPluginRegistry(t *testing.T) {
...
 		ctx: context.Background(),
 	}
 	pools, err := api.GetPools(context.Background(), &emptypb.Empty{})
 	require.NoError(t, err)
 	assert.NotEmpty(t, pools)
 	assert.NotEmpty(t, pools.AsMap())
-	assert.Equal(t, pools.AsMap()[config.Default], map[string]interface{}{"cap": 0.0, "size": 0.0})
+
+	assert.Equal(t,
+		map[string]interface{}{
```
Comment (mostafa) [inline api/api_test.go:224]: Nit: the `interface{}` can be replaced with `any` everywhere.

File: cmd/run.go [hunk 37]
```diff
@@ -622,246 +622,264 @@ var runCmd = &cobra.Command{
...
-			logger := loggers[name]
-			clientConfig := clients[name]
-			// Fill the missing and zero value with the default one.
-			cfg.HealthCheckPeriod = config.If(
-				cfg.HealthCheckPeriod > 0,
-				cfg.HealthCheckPeriod,
-				config.DefaultHealthCheckPeriod,
-			)
+		for configGroupName, configGroup := range conf.Global.Proxies {
+			for configBlcokName, cfg := range configGroup {
```
Comment (mostafa) [inline cmd/run.go:836]: Consider renaming all instances of this type:
```suggestion
			for configBlockName, cfg := range configGroup {
```

File: gatewayd.yaml (hunk 33 above)
Comment (ccoVeille) [reply gatewayd.yaml:85]: It makes sense, maybe you could/should add a comment about this where you define the constant

File: config/types.go [hunk 38]
```diff
@@ -44,19 +44,19 @@ type ActionRedisConfig struct {
...
-	ReceiveChunkSize   int           `json:"receiveChunkSize"`
-	ReceiveDeadline    time.Duration `json:"receiveDeadline" jsonschema:"oneof_type=string;integer"`
-	ReceiveTimeout     time.Duration `json:"receiveTimeout" jsonschema:"oneof_type=string;integer"`
-	SendDeadline       time.Duration `json:"sendDeadline" jsonschema:"oneof_type=string;integer"`
-	DialTimeout        time.Duration `json:"dialTimeout" jsonschema:"oneof_type=string;integer"`
-	Retries            int           `json:"retries"`
-	Backoff            time.Duration `json:"backoff" jsonschema:"oneof_type=string;integer"`
-	BackoffMultiplier  float64       `json:"backoffMultiplier"`
-	DisableBackoffCaps bool          `json:"disableBackoffCaps"`
+	Network            string        `json:"network" jsonschema:"enum=tcp,enum=udp,enum=unix" yaml:"network"`
```
Comment (mostafa) [inline config/types.go:47]: Why did you add `yaml:"..."` struct tag?

File: gatewayd.yaml [hunk 39]
```diff
@@ -32,33 +32,57 @@ metrics:
...
-    receiveDeadline: 0s # duration, 0ms/0s means no deadline
-    receiveTimeout: 0s # duration, 0ms/0s means no timeout
-    sendDeadline: 0s # duration, 0ms/0s means no deadline
-    dialTimeout: 60s # duration
-    # Retry configuration
-    retries: 3 # 0 means no retry and fail immediately on the first attempt
-    backoff: 1s # duration
-    backoffMultiplier: 2.0 # 0 means no backoff
-    disableBackoffCaps: false
+    activeWrites:
```
Comment (ccoVeille) [inline gatewayd.yaml:35]: The fact this node appears let me think (I might be wrong) this is a breaking changes.

Is it true? Ia everyone OK with that?

File: gatewayd.yaml [hunk 40]
```diff
@@ -32,33 +32,57 @@ metrics:
...
+      receiveDeadline: 0s # duration, 0ms/0s means no deadline
+      receiveTimeout: 0s # duration, 0ms/0s means no timeout
+      sendDeadline: 0s # duration, 0ms/0s means no deadline
+      dialTimeout: 60s # duration
+      # Retry configuration
+      retries: 3 # 0 means no retry and fail immediately on the first attempt
+      backoff: 1s # duration
+      backoffMultiplier: 2.0 # 0 means no backoff
+      disableBackoffCaps: false
+    standbyReads:
```
Comment (ccoVeille) [inline gatewayd.yaml:50]: I find no reference of this on the code, while I found activeWrites.

Based on the code, I would say the code will look for any node.

So, is there a need to enforce people to use activeWrites node?

Couldn't we work on a logic of weight or something like that?(maybe another PR)  So when no weight each node are equivalent

File: config/types.go (hunk 38 above)
Comment (sinadarbouy) [reply config/types.go:47]: It seems that `github.com/knadh/koanf/parsers/yaml` used in `configs.go` lowers the names in the YAML structure for nested maps if the YAML tags are not specified in the struct.

**Comment by mostafa** [comment]
The `GetServers` endpoint in the API should return the running loadbalancer config as well:
https://github.com/gatewayd-io/gatewayd/pull/577/files#diff-bf98d5fab5bcfabded9769069e7318683f549ec91cff84c5f87a9ae3a3ca3d8dR280

File: gatewayd.yaml (hunk 39 above)
Comment (mostafa) [reply gatewayd.yaml:35]: This changes is backward incompatible by design. We can opt for introducing backward compatibility, but it will increase maintenance overhead in the long run. I'll add a warning to the release notes for this.

File: api/api_test.go (hunk 36 above)
Comment (sinadarbouy) [reply api/api_test.go]: Replaced interface{} with any for all references in this PR

File: gatewayd.yaml (hunk 40 above)
Comment (sinadarbouy) [reply gatewayd.yaml:50]: The `activeWrites` and `standbyReads` can be any name. When we created the nested structure for Clients, Pools, and Proxies, we needed to set names for them. I didn’t want to use `default` because we already have a `default` for the configuration group, which would be confusing. Therefore, I used activeWrites (copied from the issue description). Now, the default name for the configuration block (nested name for Clients, Pools, Proxies) is activeWrites. That’s why you will find references to activeWrites. standbyReads was added to gateway.yaml to follow the issue structure, meaning that the gateway can support multiple Proxies(maybe its better to remove it).

File: gatewayd.yaml (hunk 33 above)
Comment (sinadarbouy) [reply gatewayd.yaml:86]: added

**Review by mostafa — APPROVED** [review]
LGTM! 🎉 

@sinadarbouy Thanks for all your efforts and contribution! 🙏 
@ccoVeille Thanks for the reviews! 🙏 
@likecodingloveproblems Thanks for the initial PR! 🙏 

I'll merge this PR and more load balancing strategies and features can be added in future PRs. We may also need to create seprate PRs for those features (and fixes).
